### ISSUE TITLE
Round metrics to the same unit, ensure some data != 0

### ISSUE THREAD

**Issue by garrett** [cockpit-project/cockpit#17769]
> - rounding: I veto "0 MB/s", as it is important to tell the difference between "idle" and "lightly busy", especially in terms of power management or figuring out if there is any active connection.

Rounding could work, as long as it's never 0 when there's actual traffic. In this case, bytes/s is basically nothing, so 0.1 MB/s is close enough in the comparison. It would, of course, require special case handling, but it's worth it, I think.

So, pseudocode would be:

```js
rounded_value = round(value, stable_unit)
if rounded == 0 then { rounded_value = round_up(value, stable_unit) }
```

Actual JS implementation, adapted from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toFixed:

```js
// Ensure that values are never 0 (unless they're supposed to be).
function round_to_string(x, decimals = 2) {
  let float_string = Number.parseFloat(x).toFixed(decimals);

  // If we get the equivalent to 0, then re-round using a ceiling
  if (!parseFloat(float_string)) {
    float_string = Number.parseFloat(
      Math.ceil(x * 10 ** decimals) / 10 ** decimals
    ).toFixed(decimals);
  }

  return float_string;
}
```

And here's the function in action:

```js
console.log(round_to_string(123.456));
// expected output: "123.46"

console.log(round_to_string('1.23e+5'));
// expected output: "123000.00"

console.log(round_to_string(0.004, 1));
// expected output: "0.1"

console.log(round_to_string(0.004, 2));
// expected output: "0.01"

console.log(round_to_string(0.004, 3));
// expected output: "0.004"

console.log(round_to_string(0.004, 0));
// expected output: "1"

console.log(round_to_string(0.000));
// expected output: "0.00"

console.log(round_to_string(0));
// expected output: "0.00"
```

Someone on the team could probably write this more elegantly. I know it could use recursion or some other tricksy thing, but I wrote it to be pretty straightforward.

> - So the only compromise that I can see is something like "< 1 MB/s" or "> 0 MB/s", or some similar notation of "a little bit, but not much". I admittedly don't like that, but I could live with that. Do you have a proposal what we should show there?

I'm also fine with this (for the less than, not for the greater than). But if we're going fractional already, then we'd still need to do special case handling. We might as well go with approximate rounding up (when rounding down returns 0) in this case and save the space (and possible confusion) of "< ".

Therefore, I prefer the rounding approach more.

> @garrett : The PF examples or body heights don't apply here. Both humans and pencils all have the same magnitude, and of course in such a situation you can get away with picking a fixed units. But network and disk speeds vary over 12 magnitudes, so it's really not the same situation.

My point is that nobody cares if it's a few bytes for one disk when something else is doing MB/s.

Of course, it's definitely also not 0 (and we shouldn't show 0 unless there really is no current transfer), but the exact number doesn't matter if it's so small and updating every second.

> If we want to change it, then the existing stuff needs to be changed

Yes, let's change it.

Sure, a compromise could be to not let it block this PR, but we'd need to do an immediate follow-up to make this PR useful. (And improve Cockpit elsewhere, especially on the metrics page.)

As a result, **I'm OK with the design** of this now, **provided we do work on making units stable in another PR**.

Note: If we're going to mention this in the release notes, the screenshot is out of date. And we'll probably want to re-take it once the stable units PR lands too.

_Originally posted by @garrett in https://github.com/cockpit-project/cockpit/pull/17657#pullrequestreview-1113511879_

**Comment by garrett** [comment]
We might want to consider adding "< " as @martinpitt suggests too. I've adjusted the demo code to handle that too, in comment https://github.com/cockpit-project/cockpit/pull/17657#issuecomment-1252190728:

Similar function, slightly modified for less-than reporting:

```js
function round_to_string(x, decimals = 2) {
  let float_string = Number.parseFloat(x).toFixed(decimals);

  // If we get a float_string equivalent to 0, then re-round using a ceiling
  // Also check to make sure x is != 0.
  if (!parseFloat(float_string) && parseFloat(x)) {
    float_string = "< " + Number.parseFloat(
      Math.ceil(x * 10 ** decimals) / 10 ** decimals
    ).toFixed(decimals);
  }

  return float_string;
}
```

Changes:

Add a check for a value in the if. Also concat a "< " in front.

**Comment by garrett** [comment]
This is a follow-up to https://github.com/cockpit-project/cockpit/pull/17657 and should be applied to the disks popup and elsewhere in metrics, especially when more than one device is shown next to another (that is: when they're being compared).

It could/should be applied to other places that show data throughput, such as Disks, Network, etc.

**Comment by garrett** [comment]
Additionally, it'd probably be useful for Cockpit-Podman and Podman-Machines too, so we probably want to have at least some of the functionality in lib and it should be sync'able.
